File: src/main/java/nz/co/aetheric/sameness/equatators/TypedEquatator.java
package nz.co.aetheric.sameness.equatators;

import nz.co.aetheric.sameness.Equatator;

/**
 * TODO: What is the purpose of this class?
 * <p>Author: <a href="http://gplus.to/tzrlk">Peter Cummuskey</a></p>
 */
@SuppressWarnings("WeakerAccess")
public abstract class TypedEquatator<T> implements Equatator<T> {

	private final Class<T> type;

	public TypedEquatator(Class<T> type) {
		this.type = type;
	}

	@Override
	public abstract <L extends T, R extends T> boolean isEquatable(L left, R right);

	public boolean isApplicable(Object other) {
		return type.isAssignableFrom(other.getClass());
	}

	public boolean isApplicable(Object left, Object right) {
		return isApplicable(left) && isApplicable(right);
	}

	public T asApplicable(Object other) {
		return type.cast(other);
	}

	public boolean attemptEquatable(Object left, Object right) {
		return isEquatable(
			asApplicable(left),
			asApplicable(right)
		);
	}

	public Class<T> getType() {
		return this.type;
	}

}

File: src/main/kotlin/nz/co/aetheric/sameness/equatators/TypedEquatator.kt
package nz.co.aetheric.sameness.equatators

import nz.co.aetheric.sameness.Equatator

/**
 * TODO: What is the purpose of this class?
 *
 * Author: [Peter Cummuskey](http://gplus.to/tzrlk)
 */
abstract class TypedEquatator<T>(override val type: Class<T>) : Equatator<T> {

	abstract override fun <L : T, R : T> isEquatable(left: L?, right: R?): Boolean

	fun isApplicable(other: Any)
			= type.isAssignableFrom(other.javaClass)

	fun isApplicable(left: Any, right: Any)
			= isApplicable(left) && isApplicable(right)

	fun asApplicable(other: Any)
			= type.cast(other)

	fun attemptEquatable(left: Any, right: Any)
			= isEquatable(asApplicable(left), asApplicable(right))

}
